File: 11-flag-variations/java/src/main/java/HostOs.java
/**
 * Host OS detection and anonymous-context evaluation for flag variations.
 * LaunchDarkly capability: Anonymous contexts + private attributes
 * See: https://launchdarkly.com/docs/sdk/features/anonymous
 * See: https://launchdarkly.com/docs/sdk/features/private-attributes
 */
public final class HostOs {
    static final String HOST_OS_ATTR = "hostOs";
    static final String ANONYMOUS_CONTEXT_KEY = "anonymous";
    static final String FLAG_ANON_OS_EMOJI = "show-anonymous-host-os-emoji";

    private HostOs() {
    }

    static String detectHostOs() {
        String osName = System.getProperty("os.name", "").toLowerCase();
        if (osName.contains("linux")) {
            return "linux";
        }
        if (osName.contains("windows")) {
            return "windows";
        }
        if (osName.contains("mac") || osName.contains("darwin")) {
            return "macos";
        }
        return "other";
    }

    static String osEmojiFor(String hostOs, boolean showOsEmoji) {
        if (!showOsEmoji) {
            return "";
        }
        return switch (hostOs) {
            case "linux" -> "🐧";
            case "macos" -> "🍎";
            case "windows" -> "🪟";
            default -> "😊";
        };
    }

    static com.launchdarkly.sdk.LDContext buildAnonymousContext(String hostOs) {
        return com.launchdarkly.sdk.LDContext.builder(ANONYMOUS_CONTEXT_KEY)
                .anonymous(true)
                .set("hostOs", hostOs)
                .privateAttributes("hostOs")
                .build();
    }

    static com.launchdarkly.sdk.LDContext buildUserContext(String username) {
        return com.launchdarkly.sdk.LDContext.builder(username).build();
    }
}

File: 10-code-control/12-flag-variations/host-os.js
/** Host OS detection and anonymous-context evaluation for flag variations. */

// LaunchDarkly capability: Anonymous contexts + private attributes
// See: https://launchdarkly.com/docs/sdk/features/anonymous
// See: https://launchdarkly.com/docs/sdk/features/private-attributes

const HOST_OS_ATTR = "hostOs";
const ANONYMOUS_CONTEXT_KEY = "anonymous";
// LaunchDarkly: flag key=show-anonymous-host-os-emoji name="Show: anonymous host OS emoji" kind=boolean
// https://app.launchdarkly.com/projects/lunch-marcoly/features/show-anonymous-host-os-emoji

const FLAG_ANON_OS_EMOJI = "show-anonymous-host-os-emoji";

const OS_EMOJI = {
  linux: "🐧",
  macos: "🍎",
  windows: "🪟",
  other: "😊",
};

function detectHostOs() {
  const platform = process.platform;
  if (platform === "linux") return "linux";
  if (platform === "win32") return "windows";
  if (platform === "darwin") return "macos";
  return "other";
}

function osEmojiFor(hostOs, showOsEmoji) {
  if (!showOsEmoji) return "";
  return OS_EMOJI[hostOs] || OS_EMOJI.other;
}

function buildAnonymousContext(hostOs) {
  return {
    kind: "user",
    key: ANONYMOUS_CONTEXT_KEY,
    anonymous: true,
    hostOs,
    privateAttributes: [HOST_OS_ATTR],
  };
}

function buildUserContext(username) {
  return { kind: "user", key: username };
}

module.exports = {
  HOST_OS_ATTR,
  ANONYMOUS_CONTEXT_KEY,
  FLAG_ANON_OS_EMOJI,
  detectHostOs,
  osEmojiFor,
  buildAnonymousContext,
  buildUserContext,
};
